src/hal/src/acceleration_structure.rs

```diff
@@ -0,0 +1,232 @@
+    pub buffer_offset: Offset,
+    pub buffer_stride: Offset,
+}
+
+// TODO doc
+#[derive(Debug)]
+pub struct AabbPositions {
+    /// Use `mint::Vector3` if available.
+    pub min: [f32; 3],
+    pub max: [f32; 3],
+}
+
+/// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkAccelerationStructureGeometryInstancesDataKHR.html
+#[derive(Debug)]
+pub struct GeometryInstances<'a, B: Backend> {
+    /// Must point to a buffer with buffer with an array of `Instance`s
+    /// TODO: document lifetime required for this buffer
+    /// TODO this struct also allows passing an array of pointers, idk if that makes sense outside the host operations case
+    pub buffer: &'a B::Buffer,
+    pub buffer_offset: Offset,
+}
+
+bitflags! {
+    /// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkGeometryInstanceFlagBitsKHR.html
+    pub struct InstanceFlags: u32 {
+        const TRIANGLE_FACING_CULL_DISABLE_BIT = 0x1;
+        const TRIANGLE_FRONT_COUNTERCLOCKWISE_BIT = 0x2;
+        const FORCE_OPAQUE_BIT = 0x4;
+        const FORCE_NO_OPAQUE_BIT = 0x8;
+    }
+}
+
+// TODO AFAIK rust doesn't have custom sized fields, so we'll need some binary writer wrapper to actually support this at an API level.
+#[derive(Debug)]
+pub struct Instance {
+    transform: TransformMatrix,
+    /// 24 bits
+    instance_custom_index: u32,
+    /// 8 bits visibility mask for the geometry. The instance may only be hit if rayMask & instance.mask != 0
+    mask: u32,
+    /// 24 bit
+    instance_shader_binding_table_record_offset: u32,
+    /// 8 bits
+    flags: InstanceFlags,
+
+    /// either B::AccelerationStructure (host operations) or GPU address (buffer + offset?)
+    acceleration_structure_reference: u64,
+}
+
+#[derive(Debug)]
+pub struct AccelerationStructureSizeRequirements {
+    pub acceleration_structure_size: u64,
+    pub update_scratch_size: u64,
+    pub build_scratch_size: u64,
+}
+
+#[derive(Debug)]
+pub enum AccelerationStructureCopyMode {
+    Clone,
+    Compact,
+
+    // these are subject to the device supporting serialization--also (at least on DX12) mainly used for debug tooling and not load perf--I'm unsure what an end-user use case looks like.
+    Serialize,
+    Deserialize,
+}
+
+#[derive(Debug)]
+pub struct AccelerationStructureBuildRangeDesc {
+    pub primitive: Range<u32>,
+    pub first_vertex: u32,
+    /// "defines an offset in bytes into the memory where a transform matrix is defined.""
+    /// TODO: this is documented as only used for `Geometry::Triangles`, but I have no idea why. GeometryTriangles::transform is implied to point to exactly one transform matrix, but I guess this pattern would allow for passing a list of transforms to reuse the same geometry (but not go through the TLAS' Instance::transform...). DX12 doesn't require any additional info at build time and thus doesn't have this concept--passing 0 here in the vulkan case would probably meet parity with DX12
```

Comment:
This is a case where my gut says we should follow DX's model of knowing the vertex/index counts upfront (during Device::create_acceleration_structure vs CommandList::build_acceleration_structure) and just not exposing vulkan's `transform_offset`. I'm open to ideas on how to implement this nicely in DX, however!

Reply:
if we go with this, would we be able to implement `gfx-portability`'s side of the KHR raytracing support?

Reply:
Hm... I guess we should support this for `gfx-portability`... I'll leave this alone for now, but we'll have to come up with a way to implement this on dx12. Maybe we can look at `vkd3d-proton` and take reverse inspiration?